### ISSUE TITLE
Support for DNS monitoring

### ISSUE THREAD

**Issue by fearful-symmetry** [elastic/ebpf#203]
### Impact

High

### Planned Version

None

### Description

So, we want Linux DNS support in endpoint, and part of that is going to be done here in ebpf. I'm making a lot of assumptions here, so feel free to correct me if there's something I'm missing. This is just a preliminary list of all the parts we'll need for DNS monitoring:

- [ ] DNS over UDP. Can be done via some combination of `ip[4,6]_datagram_connect`, `udp_destruct_sock` and others. Similar enough to existing network probes. The remaining question: do we want the probe to filter by port 53 here in ebpf, or should upstream components in endpoint do that?
- [ ] DNS over TCP. Also uses port 53. Do we want to support this?
- [ ] DNS over TLS. Uses port 853. Do we want to support his?
- [ ] DNS over HTTPS. Uses standard HTTPS port. Is there anything we can even do for this? Perhaps endpoint can just maintain a list of known DoH IP addresses to match against? Not sure if endpoint has any framework for dealing with root certs and decrypting HTTPS traffic, but that would fall outside the scope of this issue.
- [ ] Deep packet inspection. Do we care about people running DNS queries over non-standard ports? Should we start sniffing traffic over other ports looking for anything that looks like a DNS request/response?

**Comment by haesbaert** [comment]
After consulting some people, TCP/DNSSEC it's likely something we can ignore, it's not something that normal endpoint users do, maybe a security enthusiast configures on his machine and whatnot, but I think we can ignore it.

DNS@HTTPS has some momentum as firefox supports it (off by default), I doubt we could properly collect the data we want, since all we can see is the CONNECT like on the https body. I'm assuming we don't want to enter the world of uprobes (I don't).

**Comment by haesbaert** [comment]
Another point worth considering is caching local resolvers.
If the caching resolver listens to localhost:53 we're gold, but if the caching resolver can be accessed via dbus (like systemd-resolverd) or a named pipe (AF_UNIX socket), then we won't see the requests of the client hitting the local caching resolver, we would only see the outgoing request from "caching resolver->interwebs", which would be fine, the problem is it caches the reply, so further requests from the client applications would not trigger an outgoing packet, meaning we wouldn't see it.

**Comment by fearful-symmetry** [comment]
@haesbaert  yeah, I assume we're going need to care about how the local cache is set up, but unless we want to use uprobes or something for that, I assume that will fall outside the scope of changes to this repo?

**Comment by haesbaert** [comment]
> [@haesbaert](https://github.com/haesbaert) yeah, I assume we're going need to care about how the local cache is set up, but unless we want to use uprobes or something for that, I assume that will fall outside the scope of changes to this repo?

Possibly, though if it's just doing DNS over AF_UNIX we could likely add it, but I suspect glibc@nss@systemd-resolver is actually talking through dbus and thus we can't do much.
